`SpatiotemporalDispersion` estimator + more general interface for finding pixel neighbors

kahaaga opened the pull request:
# What is this PR?

This PR implement `SpatialDispersion`, a dispersion probabilities estimator for `N`-dimensional `AbstractArray`s with arbitrary stencils and symbolization schemes. The method is similar to `SpatialPermutation`, but has some additional parameters. 

This implementation builds on Azami et al. (2019), but like `SpatialSymbolicPermutation`, the implemented approach is much more flexible.

## API change
- The positional argument `p` in the code for `SpatialSymbolicPermutation` was changed to `periodic`, to reflect what it actually does. I also made it a keyword argument, so that it behaves similarly to other estimators, which most/all accept parameters as keywords.

## Additions

- [x] Implemented `stencil_length(x)`, which gives the number of possible states for an estimator based on a given stencil `x`. This is in turn used to define `alphabet_length` for new and existing spatial estimators. This addresses #122 for stencil-based symbolic estimators.
- [x] Implemented `alphabet_length(est::SpatialSymbolicPermutation)`, so that `entropy_normalized` is also defined for `SpatialSymbolicPermutation`. 

## Internal changes

- [x] Both `SpatialDispersion` and `SpatialSymbolicPermutation` are symbolic estimators, and differ only in what parameters they contain, and in the final histogram construction step. Hence, they are now in folders `symbolic/dispersion` and `symbolic/permutation_ordinal`, along with their related methods. This file/folder structure also anticipates future types of estimators that are also symbolic, but don't belong in any of the existing class of estimators.
- [x] Both `SpatialDispersion` and `SpatialSymbolicPermutation` get common functionality from `symbolic/utils_spatial.jl`, which simplifies the source files for both estimators significantly. Future estimators may also use the same machinery.

### Design decisions

- [x] Are the code structure changes above sensible?
- [x] I made an abstract type `SpatialProbEst` to allow common dispatch for `pixels_in_stencil` for any spatial probability estimator. I know we've been trying to reduce the number of intermediate abstract types, but this solution avoids having to write specialized methods for `pixels_in_stencil` for each spatial probability estimator. I'm open for other suggestions, if there are any that are more elegant/performant.

## Documentation

- [x] Some minor docs reorganization/renaming, reflecting the structure change outlined above.
- [x] Documented `SpatialDispersion`.
- [x] Added example comparing the sensitivity of normalized 2D permutation and 2D dispersion to addition of gradually stronger noise to a greyscale image.

## Tests

- Analytical tests for `SpatialDispersion`.

## References

Azami, H., da Silva, L. E. V., Omoto, A. C. M., & Humeau-Heurtier, A. (2019). Two-dimensional dispersion entropy: An information-theoretic method for irregularity analysis of images. Signal Processing: Image Communication, 75, 178-187.

ikottlarz reviewed (commented):
I @kahaaga, thanks for tagging me (and sorry for the late reply, I was sick and then on vacation)!
I'm not sure how much I can contribute here, I mainly have one point about the `alphabet_length` function, which I also implemented for `SpatialSymbolicPermutation`, but in a different way (see #131 ). Maybe we can discuss what's the best course to go here?

Datseris commented on src/probabilities_estimators/symbolic/utils_spatial.jl:
```diff
@@ -0,0 +1,80 @@
+function preprocess_spatial(stencil, x::AbstractArray, periodic::Bool = true)
+    arraysize = size(x)
+    stencil, D = stencil_to_offsets(stencil)
+    @assert length(arraysize) == D "Indices and input array must match dimensionality!"
```
Assertions are fine, but yes the alternative is 
```
length(arraysize) == D || error("...")
```
which honestly doesn't make much of a difference in neither source or what the user will see.

Datseris replied:
Where did you learn this about assertions

kahaaga replied:
> I was recently told that assertions are not good practice for these kind of checks, but should rather be implemented as Errors/ Warnings because one can ignore AssertionErrors easily. But then again, I'm not too familiar with ErrorHandling in Julia, so maybe this is different here?
> Where did you learn this about assertions

[This](https://discourse.julialang.org/t/efficient-use-of-test-or-assert/75895/6) answer on JuliaDiscourse sums it up nicely: assertions in Julia can be turned off as an optimisation, so it shouldn't be user for input checking or similar things (which is what we're doing here).

> which honestly doesn't make much of a difference in neither source or what the user will see

```julia
length(arraysize) == D || error("...")
```

Using `error()` here will throw the most generic error possible, i.e. `ErrorException`. I think it is good practice to throw  error types that accurately reflect what type of condition was tested. This takes away the need for long error messages, and makes tests easier to write/read/understand. There's probably a build-in error that handles checks like these reasonably well.

Alternatively, one could write a custom error type like so:

```julia
type InputDimensionError{S<:AbstractString} <: Exception
    msg::S
end
Base.showerror(io::IO, e::InputDimensionError) = print(io, "InputDimensionError occured because ", e.msg, ".")
```
which could then be called as 

```julia
length(arraysize) == D || throw(InputDimensionError("Indices and input array must match dimensionality!")
```

This is arguably over-the-top for a simple input check like this, but can in principle be done.

Datseris replied:
Standard error is fine. The most important infromation is in the string the error throws. In any case here the `DimensionMissmatch` error fits the most.

Datseris commented on src/probabilities_estimators/symbolic/dispersion/spatial_dispersion.jl:
```diff
@@ -0,0 +1,132 @@
+import Base.maximum
+
+"""
+    SpatialDispersion <: ProbabilitiesEstimator
```
No, this is the documentation string. The information exposed to the user should be `<: ProbabilitiesEstimator` because that's what the user cares about.

kahaaga replied:
> This should be a SpatialProbEst{D, P}, right?

As detailed in my first comment to this PR, `SpatialProbEst{D, P}` type is an *implementation detail* introduced only to simplify the codebase, because we now have multiple spatial estimators. This internal structure of the code might change without it ever affecting the end user. 

Without it being stated anywhere in the code or docs for Entropies.jl (woops, maybe be should :D), we follow Julia's version of [https://semver.org/](semantic versioning). Point 1 in semver says that a package must declare a *public API*. The public API is *strictly* what exists in the documentation.

We don't want implementation details to unnecessarily clutter the public API, so it is best to only expose what is needed. `SpatialDispersion` *is* a `SpatialProbEst{D, P}`, but as @Datseris said, that is not relevant to the user.  Documenting `SpatialDispersion <: ProbabilitiesEstimator` is sufficient. Hypothetically, in the future, we might even end up removing `SpatialProbEst` completely. As long as it's not documented, internal changes will have zero effects on the end user.

Sidenote: the name `SpatialDispersion` itself implies that this is a spatial estimator too, so this information is already there implicitly.

kahaaga commented on src/probabilities_estimators/symbolic/dispersion/spatial_dispersion.jl:
```diff
@@ -0,0 +1,132 @@
+    print(io, "Stencil: ")
+    show(io, MIME"text/plain"(), est.stencil)
+    print(io, "\nSymbolization: $(est.symbolization)")
+    print(io, """\nBoundaries: $(est.periodic ? "Periodic" : "Non-periodic")""")
```
Well spotted. Whether the estimator is periodic or not is actually inferred from the type,`SpatialDispersion{D,P,V,S}`, where by constructor construction (:D), the second type parameter is a boolean indicating whether or not the estimator is periodic. I fixed this this in my latest commit.

kahaaga commented on src/probabilities_estimators/symbolic/utils_spatial.jl:
```diff
@@ -0,0 +1,80 @@
+function stencil_length end
+stencil_length(stencil::Vector{CartesianIndex{D}}) where D = length(stencil)
+stencil_length(stencil::Array{Int, D}) where D = sum(stencil)
+stencil_length(stencil::NTuple{2, NTuple{D, T}}) where {D, T} = factorial(prod(stencil[1]))
```
As mentioned elsewhere, I think having `m` as a field in `SpatialSymbolicPermutation` is fine.

kahaaga commented:
I'm assigning this to the 2.0 milestone, because we need a slightly more generic interface, like I propose `utils_spatial.jl`, to be able to generically compute any entropy on N-dimensional grids. We will also be able to utilize this machinery directly in Complexity.jl. 

I just have to update the code once we've merged the other outstanding PRs.

kahaaga commented:
Hey @Datseris and @ikottlarz! A lot has changed since I initially submitted this PR. I've now updated the PR according to these changes.

In short, these are the current changes relative to the `main` branch:

- Introduced the `SpatialDispersion` estimator. The 2D case of spatial dispersion entropy was introduced in Azami et al. (2019). Here  it is defined for `N`-dimensional data (using our brilliant spatial permutation interface).
- `m` is now a field in both `SpatialSymbolicPermutation` and `SymbolicDispersion` (as introduced by @ikottlarz in #131).
- Changed the file structure a bit, so that all spatial estimators is in one folder, with a subfolder structure that mimics  the `src/probabilities_estimators/` folder. I did the same for the tests.
- The `utils.jl` and `spatial.jl` files `src/probabilities_estimators/spatial` defines common functionality for all existing and future spatial estimators. The `SpatialProbEst <: ProbabilitiesEstimator` supertype makes dispatch easier, and is only used internally.
- The spatial estimators are quite complicated. Upon revisiting the code after som time away, I had time spend quite some time to understand how to make the different stencils.  Therefore, I slightly expanded the "usage" sections, at the cost of being slightly more verbose, but with the benefit that the user immediately understand how to use the estimator. 
- Analytical tests for both `SpatialSymbolicPermutation` and `SymbolicDispersion` are included.
- The performance of `SpatialSymbolicPermutation` and `SymbolicDispersion` is demonstrated on a 2D image with progressively more noise added.

## References

Azami, H., da Silva, L. E. V., Omoto, A. C. M., & Humeau-Heurtier, A. (2019). Two-dimensional dispersion entropy: An information-theoretic method for irregularity analysis of images. Signal Processing: Image Communication, 75, 178-187.

## What remains before merging?

This PR is one of the last pieces remaining before we release 2.0, so I'd like to request a re-review on this PR from at least one of you before we merge this. 

I know you're both very busy at the moment, @Datseris and @ikottlarz, so please don't stress about it. In the meantime, I can just check out the `spatiotemporal_dispersion` branch whenever I need to use this locally.

kahaaga commented:
To avoid dealing with too many messy merge conflicts, I'm closing this in favor of a new PR, which uses the updated encoding API.
